framework/include/userobject/InterfaceQpMaterialPropertyBaseUserObject.h

```diff
@@ -0,0 +1,113 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "InterfaceQpUserObjectBase.h"
+
+// Forward declarations
+template <typename T = Real>
+class InterfaceQpMaterialPropertyBaseUserObject;
+
+template <>
+InputParameters validParams<InterfaceQpMaterialPropertyBaseUserObject<>>();
+
+/**
+ * Specialization of InterfaceQpUserObjectBase for material properties. Material property type
+ * specialization is achieved by specializing computeScalarMaterialProperty in derived classes.
+ */
+template <typename T>
+class InterfaceQpMaterialPropertyBaseUserObject : public InterfaceQpUserObjectBase
+{
+
+public:
+  static InputParameters validParams();
+  /**
+   * Class constructor
+   * @param parameters The input parameters for this object
+   */
+  InterfaceQpMaterialPropertyBaseUserObject(const InputParameters & parameters);
+  virtual ~InterfaceQpMaterialPropertyBaseUserObject(){};
+
+protected:
+  /* method computing the real value*/
+  virtual Real computeRealValue(const unsigned int qp) override final
```

Comment:
I would prefer that the definition be outside the class body. Its best for developers trying to get a grasp of a class API to scroll through as few lines as possible. If (longer than one or two line) definitions are included in the class body, it takes much longer to scroll

Reply:
I moved all definitions to the C files